eth/db/diff.py

```diff
@@ -53,72 +66,74 @@ class DBDiffTracker(MutableMapping):
     When it's time to take the tracked changes and write them to your database,
     get the :class:`DBDiff` with :meth:`DBDiffTracker.diff` and use the attached methods.
     """
-    def __init__(self):
-        self._changes = {}  # type: Dict[bytes, Union[bytes, DiffMissingError]]
+    def __init__(self) -> None:
+        self._changes = {}  # type: Dict[bytes, Union[bytes, MissingReason]]
 
-    def __contains__(self, key):
+    def __contains__(self, key: bytes) -> bool:     # type: ignore # Breaks LSP
         result = self._changes.get(key, NEVER_INSERTED)
         return result not in (DELETED, NEVER_INSERTED)
 
-    def __getitem__(self, key):
+    def __getitem__(self, key: bytes) -> bytes:
         result = self._changes.get(key, NEVER_INSERTED)
         if result in (DELETED, NEVER_INSERTED):
-            raise DiffMissingError(key, result)
+            raise DiffMissingError(key, cast(MissingReason, result))
         else:
-            return result
+            return cast(bytes, result)
```

Comment:
@pipermerriam @carver I'm worried about the usage of `cast` in performance sensitive code paths.

Here's @gvanrossum from 2016

>I would really hope we won't need to use or encourage casts here, since it's an expensive runtime operation (at least it is until we teach CPython about it -- right now it is a user-defined function, which is incredibly slow compared to a plain assignment)

I'm not sure if anything has changed about this in the meantime but if not then I assume that even if `cast` is a `noop` at runtime it is still one extra function call and hence will slow things down. Therefore I'd advocate to use `type: ignore` in such performance sensitive code paths. Thoughts?

Reply:
If the best solution we have is a cast, then yes, a `type: ignore` seems like a fine alternative to me. (given that it's in a performance-sensitive line, and the line is doing exactly one thing -- we don't want to accidentally ignore other type checks on the same line).